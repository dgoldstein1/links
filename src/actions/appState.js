import { store } from "../reducers/index";
import { fetchAndStoreRandomStartNode } from "./graph";
import { postUserVisit } from "../api/analytics";

// updates which view the app is in
export const UPDATE_VIEW = "UPDATE_VIEW";
export function updateView(view) {
  return {
    type: UPDATE_VIEW,
    view: view
  };
}
export const SET_LOADING = "SET_LOADING";
export function setLoading(loading) {
  return {
    type: SET_LOADING,
    loading: loading
  };
}
export const SET_FATAL_ERROR = "SET_FATAL_ERROR";
export function setFatalError(e) {
  return {
    type: SET_FATAL_ERROR,
    fatalError: e
  };
}

export const SET_LANGUAGE = "SET_LANGUAGE";
export function setLanguage(language) {
  return {
    type: SET_LANGUAGE,
    language
  };
}

// initializes app on first load
export function InitAapp(callback = () => {}) {
  store.dispatch(setLoading(true));
  // send info to analytics backend
  postUserVisit();

  fetchAndStoreRandomStartNode(err => {
    if (err) store.dispatch(setFatalError(err));
    store.dispatch(setLoading(false));
    callback();
  });
}
