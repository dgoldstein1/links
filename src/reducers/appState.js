import { SET_LOADING, UPDATE_VIEW, SET_FATAL_ERROR } from "../actions/appState";

const initialState = {
  loading: true, // the app is / isn't loading
  fatalError: undefined, // error which  causes app to not load
  view: "splash" // one of "explore", "path", "splash", "about", "settings"
};

const visitReducer = (state = initialState, action) => {
  switch (action.type) {
    case SET_FATAL_ERROR:
      return Object.assign({}, state, {
        fatalError: action.fatalError
      });
    // set when the app is loading
    case SET_LOADING:
      return Object.assign({}, state, {
        loading: action.loading === undefined ? false : action.loading
      });
    // update where the user is currently looking
    case UPDATE_VIEW:
      return Object.assign({}, state, {
        view: action.view
      });
    default:
      return state;
  }
};

export default visitReducer;
