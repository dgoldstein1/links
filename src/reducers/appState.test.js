import appState from "./appState";

describe("reducers", () => {
  describe("appState", () => {
    const initialState = {
      loading: true, // the app is / isn't loading
      view: "splash",
    };

    it("initializes with correct state", () => {
      expect(appState(undefined, { action: undefined })).toMatchSnapshot();
    });

    describe("UPDATE_VIEW", () => {
      it("updates the store with a new view", () => {
        let action = {
          type: "UPDATE_VIEW",
          view: "newView",
        };
        expect(appState(undefined, action)).toMatchSnapshot();
      });
    });
    describe("SET_LOADING", () => {
      it("updates the store with new value of loading", () => {
        let action = {
          type: "SET_LOADING",
          loading: true,
        };
        expect(appState(undefined, action)).toMatchSnapshot();
      });
    });
    describe("SET_LANGUAGE", () => {
      it("sets new language", () => {
        expect(
          appState(undefined, { type: "SET_LANGUAGE", language: "newlang" })
        ).toMatchSnapshot();
      });
    });
  });
});
