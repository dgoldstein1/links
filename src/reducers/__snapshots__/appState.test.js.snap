// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reducers appState SET_LOADING updates the store with new value of loading 1`] = `
Object {
  "fatalError": undefined,
  "language": "english",
  "loading": true,
  "view": "splash",
}
`;

exports[`reducers appState UPDATE_VIEW updates the store with a new view 1`] = `
Object {
  "fatalError": undefined,
  "language": "english",
  "loading": true,
  "view": "newView",
}
`;

exports[`reducers appState initializes with correct state 1`] = `
Object {
  "fatalError": undefined,
  "language": "english",
  "loading": true,
  "view": "splash",
}
`;
